test/common/filter/ext_authz_test.cc

```diff
@@ -0,0 +1,301 @@
+
+  EXPECT_EQ(Network::FilterStatus::Continue, filter_->onData(data));
+
+  EXPECT_CALL(*client_, cancel()).Times(0);
+  filter_callbacks_.connection_.raiseEvent(Network::ConnectionEvent::LocalClose);
+
+  EXPECT_EQ(1U, stats_store_.counter("ext_authz.name.total").value());
+  EXPECT_EQ(1U, stats_store_.counter("ext_authz.name.ok").value());
+}
+
+TEST_F(ExtAuthzFilterTest, DeniedWithSSLConnect) {
+  InSequence s;
+
+  EXPECT_CALL(filter_callbacks_.connection_, remoteAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(filter_callbacks_.connection_, localAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(filter_callbacks_.connection_, readDisable(true));
+  EXPECT_CALL(*client_, check(_, _, _))
+      .WillOnce(WithArgs<0>(
+          Invoke([&](RequestCallbacks& callbacks) -> void { request_callbacks_ = &callbacks; })));
+
+  EXPECT_EQ(Network::FilterStatus::Continue, filter_->onNewConnection());
+  // Called by SSL when the handshake is done.
+  filter_callbacks_.connection_.raiseEvent(Network::ConnectionEvent::Connected);
+  Buffer::OwnedImpl data("hello");
+  EXPECT_EQ(Network::FilterStatus::StopIteration, filter_->onData(data));
+
+  EXPECT_CALL(filter_callbacks_.connection_, readDisable(false));
+  EXPECT_CALL(filter_callbacks_.connection_, close(Network::ConnectionCloseType::NoFlush));
+  EXPECT_CALL(*client_, cancel()).Times(0);
+  request_callbacks_->onComplete(CheckStatus::Denied);
+
+  EXPECT_EQ(Network::FilterStatus::Continue, filter_->onData(data));
+
+  EXPECT_EQ(1U, stats_store_.counter("ext_authz.name.total").value());
+  EXPECT_EQ(1U, stats_store_.counter("ext_authz.name.unauthz").value());
+  EXPECT_EQ(1U, stats_store_.counter("ext_authz.name.cx_closed").value());
+}
+
+TEST_F(ExtAuthzFilterTest, FailOpen) {
+  InSequence s;
+
+  EXPECT_CALL(filter_callbacks_.connection_, remoteAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(filter_callbacks_.connection_, localAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(*client_, check(_, _, _))
+      .WillOnce(WithArgs<0>(
+          Invoke([&](RequestCallbacks& callbacks) -> void { request_callbacks_ = &callbacks; })));
+
+  EXPECT_EQ(Network::FilterStatus::Continue, filter_->onNewConnection());
+  Buffer::OwnedImpl data("hello");
+  EXPECT_EQ(Network::FilterStatus::StopIteration, filter_->onData(data));
+
+  EXPECT_CALL(filter_callbacks_.connection_, close(_)).Times(0);
+  EXPECT_CALL(*client_, cancel()).Times(0);
+  EXPECT_CALL(filter_callbacks_, continueReading());
+  request_callbacks_->onComplete(CheckStatus::Error);
+
+  EXPECT_EQ(Network::FilterStatus::Continue, filter_->onData(data));
+
+  EXPECT_EQ(1U, stats_store_.counter("ext_authz.name.total").value());
+  EXPECT_EQ(1U, stats_store_.counter("ext_authz.name.error").value());
+  EXPECT_EQ(0U, stats_store_.counter("ext_authz.name.unauthz").value());
+  EXPECT_EQ(0U, stats_store_.counter("ext_authz.name.cx_closed").value());
+}
+
+TEST_F(ExtAuthzFilterTest, Error) {
```

Comment:
Not a very descriptive test name. And I am actually having a tough time groking what it is that you want to test here. Do you mind clarifying

Reply:
I also like the idea of adding a leading comment line for tests that aren't obvious/trivial. E.g.:

```
// Validate error handling when foo does bar with baz.
TEST_F(ExtAuthzFilterTest, Error) {
```